### PR TITLE
[SPARK-46908][SQL] Support star clause in WHERE clause

### PR DESCRIPTION
**Pull request by srielau**
### What changes were proposed in this pull request?

- Extend star clause from the Project to Filter
- Treat (*) as a row value constructor, rather than a no-op
- Support IN(*)

### Why are the changes needed?

This PR simplifies the writing of queries by bringing the popular shorthand to more scenarios.
For example one can write:
WHERE coalesce(* EXCEPT(pk)) IS NOT NULL to remove "null rows") or
WHERE foo(t.*) = true to pass all columns from a table to a function
 
### Does this PR introduce _any_ user-facing change?

Yes

### How was this patch tested?

New tests in selectExcept

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by gatorsmile** [comment]
@gengliangwang @cloud-fan

File: docs/sql-ref-function-invocation.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+table_ordering
+  { { ORDER | SORT } BY { order_by_expr | ( order_by_expr [, ...] } }
+```
+
+### Parameters
+
+- **function_name**
+
+  The name of the built-in or user defined function. When resolving an unqualified function_name Databricks will first consider a built-in or temporary function, and then a function in the current schema.
```
Comment (cloud-fan) [inline docs/sql-ref-function-invocation.md:66]:
```suggestion
  The name of the built-in or user defined function. When resolving an unqualified function_name Spark will first consider a built-in or temporary function, and then a function in the current schema.
```

File: docs/sql-ref-function-invocation.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    Identifies a table to pass to the function by name.
+
+  - **TABLE ( query )**
+
+    Passes the result of query to the function.
+
+  - **table-partitioning**
+
+    Optionally specifies that the table argument is partitioned. If not specified the partitioning is determined by Databricks.
```
Comment (cloud-fan) [inline docs/sql-ref-function-invocation.md:94]:
```suggestion
    Optionally specifies that the table argument is partitioned. If not specified the partitioning is determined by Spark.
```

File: docs/sql-ref-syntax-qry-select-star.md
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select-star.md:1]: since it's not limited to select clause now, shall we just say `sql-ref-syntax-qry-star`?

File: docs/sql-ref-syntax-qry-select-star.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+### Description
+
+A shorthand to name all the referencable columns in the FROM clause or a specific table reference in the FROM clause.
+The star clause clause is most frequently used in the SELECT list.
+Spark also supports its use in function invocation and certain n-ary operations within the SELECT list and WHERE clause.
+
+### Syntax
+
+```
+[ { table_name | view_name } . ] * [ except_clause ]
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select-star.md:31]: it's actually more than it, as it can be a column or inner field, e.g. `t1.col.*` and `t1.col.innerField.*`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -1534,6 +1534,9 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
       // If the projection list contains Stars, expand it.
       case p: Project if containsStar(p.projectList) =>
         p.copy(projectList = buildExpandedProjectList(p.projectList, p.child))
+      // If the filter list contains Stars, expand it.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1537]: shall we apply it for all operators?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -1936,6 +1939,11 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
             case o => o :: Nil
           })
         // count(*) has been replaced by count(1)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1941]: shall we keep this comment above the last default case match?

File: sql/core/src/test/resources/sql-tests/analyzer-results/subquery/in-subquery/in-group-by.sql.out [hunk 6]
```diff
@@ -1,4 +1,4 @@
--- Automatically generated by SQLQueryTestSuite
+git commit -- Automatically generated by SQLQueryTestSuite
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/analyzer-results/subquery/in-subquery/in-group-by.sql.out:1]: unnecessary change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -1864,8 +1867,8 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
      * Build a project list for Project/Aggregate and expand the star if possible
      */
     private def buildExpandedProjectList(
-      exprs: Seq[NamedExpression],
-      child: LogicalPlan): Seq[NamedExpression] = {
+                                          exprs: Seq[NamedExpression],
+                                          child: LogicalPlan): Seq[NamedExpression] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1871]: unnecessary change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 4 above)
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1537]: Define all. I don't think it's useful fro GROUP BY (ALL is better), and semantic is not clear (or dangerous) for ORDER BY.

File: docs/sql-ref-syntax-qry-select-star.md (hunk 3 above)
Comment (srielau) [reply docs/sql-ref-syntax-qry-select-star.md:31]: Seriously? It unnests fields??? People have been asking for that....

File: sql/core/src/test/resources/sql-tests/analyzer-results/subquery/in-subquery/in-group-by.sql.out (hunk 6 above)
Comment (srielau) [reply sql/core/src/test/resources/sql-tests/analyzer-results/subquery/in-subquery/in-group-by.sql.out:1]:
```suggestion
-- Automatically generated by SQLQueryTestSuite
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 7 above)
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1871]:
```suggestion
      child: LogicalPlan): Seq[NamedExpression] = {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -1864,7 +1867,7 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
      * Build a project list for Project/Aggregate and expand the star if possible
      */
     private def buildExpandedProjectList(
-      exprs: Seq[NamedExpression],
+                                          exprs: Seq[NamedExpression],
```
Comment (srielau) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1870]:
```suggestion
      exprs: Seq[NamedExpression],
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1537]: oh makes sense

File: docs/sql-ref-syntax-qry-select-star.md
Comment (srielau) [reply docs/sql-ref-syntax-qry-select-star.md:1]: Done

File: docs/sql-ref-syntax-qry-select-star.md (hunk 3 above)
Comment (srielau) [reply docs/sql-ref-syntax-qry-select-star.md]: Doc-ed and added a test. Are there any other secrets in Spark I should know about?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 9]
```diff
@@ -1936,6 +1939,11 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
             case o => o :: Nil
           })
         // count(*) has been replaced by count(1)
+        case p: In if containsStar(p.children) =>
+          p.copy(list = p.list.flatMap {
+            case s: Star => expand(s, child)
+            case o => o :: Nil
+          })
```
Comment (srielau) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1946]:
```suggestion
          })
          // count(*) has been replaced by count(1)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 5 above)
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1941]:
```suggestion
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 10]
```diff
@@ -1935,7 +1938,12 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
             case s: Star => expand(s, child)
             case o => o :: Nil
           })
-        // count(*) has been replaced by count(1)
+        case p: In if containsStar(p.children) =>
+          p.copy(list = p.list.flatMap {
+            case s: Star => expand(s, child)
+            case o => o :: Nil
+          })
+          // count(*) has been replaced by count(1)
```
Comment (srielau) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1946]:
```suggestion
        // count(*) has been replaced by count(1)
```

File: docs/sql-ref-syntax-qry-star.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+
+### Description
+
+A shorthand to name all the referencable columns in the FROM clause or a specific table reference's columns or fields in the FROM clause.
+The star clause clause is most frequently used in the SELECT list.
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-star.md:25]:
```suggestion
The star clause is most frequently used in the SELECT list.
```

File: docs/sql-ref-syntax-qry-star.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+except_clause
+   EXCEPT ( { column_name | field_name } [, ...] )
+```
+
+### Parameters
+
+* **name**
+
+  If present limits the columns or fields to be named to those in the specified referencable field, columns, or table.
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-star.md:41]:
```suggestion
  If present limits the columns or fields to be named to those in the specified referencable field, column, or table.
```

File: docs/sql-ref-syntax-qry-star.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+* **`*`**
+
+  Collects all the referencable columns in the FROM clause or the optionally specified table_name or view_name into a column list.
+  The list of columns is ordered by the order of table_references and the order of columns within each table_reference.
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-star.md:46]: This seems duplicated with the overall description, shall we remove it?

**Review by cloud-fan — APPROVED** [review]

File: docs/sql-ref-syntax-qry-star.md (hunk 13 above)
Comment (srielau) [reply docs/sql-ref-syntax-qry-star.md]: Agreed

**Comment by srielau** [comment]
All comments addressed. This PR is ready to merge once it's done with QA cycle.
@cloud-fan

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
@srielau +1 for the proposal! It would be great if we could mention the change in https://github.com/apache/spark/blob/master/docs/sql-migration-guide.md as well.

**Comment by srielau** [comment]
> @srielau +1 for the proposal! It would be great if we could mention the change in https://github.com/apache/spark/blob/master/docs/sql-migration-guide.md as well.

Done @gengliangwang

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
